### ISSUE TITLE
uninitialized constant Wice::ViewColumnInteger

### ISSUE THREAD

**Issue by creativetags** [leikind/wice_grid#74]
Just installed wice_grid (3.2.0.pre1) on rails 3.2.6 and got the error:

```
gems/wice_grid-3.2.0.pre1/lib/view_columns/view_column_float.rb:4: uninitialized constant Wice::ViewColumnInteger (NameError)
```

whilst running rails g wice_grid:install

**Comment by leikind** [comment]
That's weird. I've just created an empty 3.2.6 app, added wice_grid to the gemfile , run bundle, and then rails g wice_grid:install. No errors.

Please submit the full stacktrace.

Does `rails c` work?

What is in the gemfile?

**Comment by creativetags** [comment]
rails c gives the same:
/Users/mark/.rvm/gems/ree-1.8.7-2012.02@Example/gems/wice_grid-3.2.0.pre1/lib/view_columns/view_column_float.rb:4: uninitialized constant Wice::ViewColumnInteger (NameError)
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@Example/gems/wice_grid-3.2.0.pre1/lib/view_columns.rb:12:in `load_column_processors'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@Example/gems/wice_grid-3.2.0.pre1/lib/view_columns.rb:10:in`each'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@Example/gems/wice_grid-3.2.0.pre1/lib/view_columns.rb:10:in `load_column_processors'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@Example/gems/wice_grid-3.2.0.pre1/lib/wice_grid.rb:49
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/activesupport-3.2.6/lib/active_support/lazy_load_hooks.rb:36:in`instance_eval'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/activesupport-3.2.6/lib/active_support/lazy_load_hooks.rb:36:in `execute_hook'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/activesupport-3.2.6/lib/active_support/lazy_load_hooks.rb:26:in`on_load'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/activesupport-3.2.6/lib/active_support/lazy_load_hooks.rb:25:in `each'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/activesupport-3.2.6/lib/active_support/lazy_load_hooks.rb:25:in`on_load'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@Example/gems/wice_grid-3.2.0.pre1/lib/wice_grid.rb:39
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/initializable.rb:30:in `instance_exec'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/initializable.rb:30:in`run'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/initializable.rb:55:in `run_initializers'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/initializable.rb:54:in`each'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/initializable.rb:54:in `run_initializers'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/application.rb:136:in`initialize!'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/railtie/configurable.rb:30:in `send'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/railtie/configurable.rb:30:in`method_missing'
    from /Users/mark/Example/config/environment.rb:5
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/application.rb:103:in `require_environment!'
    from /Users/mark/.rvm/gems/ree-1.8.7-2012.02@global/gems/railties-3.2.6/lib/rails/commands.rb:40
    from script/rails:6:in`require'
    from script/rails:6

**Comment by leikind** [comment]
Does the problem persist with 3.2.0.pre2 ?

**Comment by creativetags** [comment]
Same error with 3.2.0.pre2

**Comment by leikind** [comment]
I think I know the cause of the bug . 

I see you are using ruby 1.8.7 (ree 1.8.7).  Hashes are ordered in 1.9. That is, iterating this hash https://github.com/leikind/wice_grid/blob/rails3/lib/view_columns/column_processor_index.rb in 1.9 will follow the same order as in the definition. Not in 1.8.7. And the order is essential.

This will be fixed in 3.2.0.pre3. 

Meanwhile, if you can't wait, you could either switch to 1.9.3, or hack the plugin to use ActiveSupport::OrderedHash, I can explain how

**Comment by creativetags** [comment]
Thanks, 1.9.3 isn't an option for me right now. I'd appreciate help switching to orderedhash.

**Comment by leikind** [comment]
Ok. I think I will release the fix in a few days, meanwhile, that's how you do this:

1) Copy the contents of your gem to vendor/wice_grid. This can be done by a command like 

```
cp -a `bundle show wice_grid`   vendor/wice_grid
```

provided your current directory is your app

2) gem uninstall  wice_grid (just to be sure)

3) In Gemfile replace

```
gem "wice_grid",  '3.2.0.pre2'
```

by 

```
gem "wice_grid",  :path => 'vendor/wice_grid'
```

4) run

```
 bundle
```

5) Now that you have your own version of WiceGrid replace

```
COLUMN_PROCESSOR_INDEX = {
  :action     => 'action_view_column', # Special processor for action column, columns with checkboxes
  :text       => 'view_column_string',
  :string     => 'view_column_string',
  :timestamp  => 'view_column_datetime',
  :datetime   => 'view_column_datetime',
  :date       => 'view_column_date',
  :integer    => 'view_column_integer',
  :float      => 'view_column_float',
  :decimal    => 'view_column_float',
  :custom     => 'view_column_custom_dropdown',  # Special processor for custom filter columns
  :boolean    => 'view_column_boolean'
}
```

in wice_grid/lib/view_columns/column_processor_index.rb by

```
COLUMN_PROCESSOR_INDEX = ActiveSupport::OrderedHash[
  :action   , 'action_view_column', # Special processor for action column, columns with checkboxes
  :text     , 'view_column_string',
  :string   , 'view_column_string',
  :timestamp, 'view_column_datetime',
  :datetime , 'view_column_datetime',
  :date     , 'view_column_date',
  :integer  , 'view_column_integer',
  :float    , 'view_column_float',
  :decimal  , 'view_column_float',
  :custom   , 'view_column_custom_dropdown',  # Special processor for custom filter columns
  :boolean  , 'view_column_boolean'
]
```

This should fix the problem.

If it works, feel free to fork the project, commit the change and submit the pull request

**Comment by creativetags** [comment]
Thanks, that worked

**Comment by creativetags** [comment]
I don't know if this is related but I'm getting 

```
undefined method `filter_all_label=' for #<Wice::ViewColumn:0x10ab244a8>
```

around the 2nd line here:

```
= grid(@users_grid, :show_filters => :no) do |g|
    = g.column do |user|
        = user.company
```

**Comment by creativetags** [comment]
Ok, I got past that but now I'm getting 

```
undefined method `per' for #<ActiveRecord::Relation:0x103e01470>
```

called on line 299 of wice_grid.rb

**Comment by leikind** [comment]
Add 

```
 gem 'kaminari', '0.13.0'
```

to Gemfile and run bundle. Normally kaminari is installed as a dependancy by Bundler, but in this case this is probably not happening because you are using the :path declaration.

**Comment by creativetags** [comment]
That didn't solve the problem unfortunately.

**Comment by leikind** [comment]
could you  verify that kaminari is in `Gemfile.lock` (not `Gemfile`)

**Comment by creativetags** [comment]
It's the same version 0.13.0

**Comment by leikind** [comment]
and if you run `rails c` and type `Kaminari`, is there an exception. If there isn't, can you try to paginate some active records with kaminari to verify that it is loaded:

```
 YourARModel.page(7).per(50)
```

**Comment by creativetags** [comment]
No exception when I run Kaminari on console and pagination works but still getting that error undefined method `per' in my app.

**Comment by leikind** [comment]
My theory is that you still have `will_paginate` loaded and `page` from `will_paginate` gets shadowed by `page` from `will_paginate` https://github.com/mislav/will_paginate/blob/master/lib/will_paginate/active_record.rb#L140

Do you have `will_paginate` in `Gemfile.lock`? If you only have it there for WiceGrid, get rid of it. If you use it, remove it at least temporarily to prove that my theory is correct.

**Comment by leikind** [comment]
Oops, I meant:
`page` from `kaminari` gets shadowed by page from `will_paginate`

**Comment by creativetags** [comment]
Uninstalling will_paginate fixed it! Thanks for all your help.

**Comment by leikind** [comment]
Thank you for all your precious feedback. I will add a warning to the documentation saying that since 3.2.0.pre2 WiceGrid is not compatible with will_paginate because kaminari is not compatible with will_paginate.

**Comment by creativetags** [comment]
Another question for you - do you know how to convert the view to HAML?

```
<%= grid(@tasks_grid) do |g|

  g.column do |task|
    task.id
  end

  g.column  do |task|
    task.title
  end

  g.column do |task|
    task.description
  end

  g.column do |task|
    task.archived? ? 'Yes' : 'No'
  end

  g.column do |task|
    link_to('Edit', edit_task_path(task))
  end
end -%>
```

**Comment by leikind** [comment]
I love HAML and I experimented converting it to HAML, I cannot say now how exactly  I did it,  but it looked really ugly and unreadable , so I decided for myself that the views with such grid definitions should stay in ERB. Normally I have a view per grid definition, this is handy and a requirement if you need csv export. But if you think about it, a grid definition is actually a method call with a block and more code inside,  a chunk of Ruby code, i.e. a helper, so another approach is to put them into helpers thus keeping your views HAML-only

**Comment by creativetags** [comment]
Thanks, that's helpful. I got round it by using the :erb filter to embed the erb code in my haml file.

**Comment by leikind** [comment]
3.2.0.pre3 has been released with the fix for ruby 1.8.7

**Comment by creativetags** [comment]
Thanks, I've switched to it and it's working well.
